kernel/src/sched.rs

```diff
@@ -332,31 +424,39 @@ impl Kernel {
         }
     }
 
-    /// Main loop.
-    pub fn kernel_loop<P: Platform, C: Chip>(
-        &'static self,
+    /// Main loop of the OS
+    pub fn kernel_loop<P: Platform, C: Chip, SC: Scheduler<C>>(
+        &self,
         platform: &P,
         chip: &C,
         ipc: Option<&ipc::IPC>,
+        scheduler: &SC,
         _capability: &dyn capabilities::MainLoopCapability,
-    ) {
-        chip.watchdog().setup();
+    ) -> ! {
         loop {
```

Comment:
I am having a hard time reasoning about if this loop can be constructed according to my other comments.

Reply:
I think the solution is to not have a nested loop. The nested loop isn't really needed and makes reasoning about program flow a lot more confusing.